File: apple/Nianxiang/UI/GraphLayout.swift
import Foundation

/// Force layout simulation for the relationship graph. Pure port of the web
/// GraphOverlay.tsx `useForceLayout` step: same constants, same three passes
/// (pairwise repulsion → edge springs → center pull + wander + damping + integrate).
enum GraphLayout {
    static let repulsion: Double = 2500
    static let springK: Double = 0.04
    static let springRestAI: Double = 90
    static let springRestCooccur: Double = 140
    static let centerK: Double = 0.004
    static let damping: Double = 0.9
    static let wander: Double = 0.035

    struct Point: Equatable {
        var x: Double
        var y: Double
        var vx: Double = 0
        var vy: Double = 0
        var fixed = false
        var phase: Double = 0
    }

    struct Edge {
        let a: String
        let b: String
        /// true when the edge carries a label (AI/manual relation); false = co-occurrence only.
        /// Selects the spring's rest length, mirroring `e.label ? SPRING_REST_AI : SPRING_REST_COOCCUR`.
        let hasLabel: Bool
    }

    /// Initial placement for a freshly-seen node: `0.35 * min(w,h)` radius circle, positioned
    /// by index, with a per-id hashed phase for the idle wander drift.
    static func initialPoint(id: String, index: Int, count: Int, width: Double, height: Double) -> Point {
        let cx = width > 0 ? width / 2 : 300
        let cy = height > 0 ? height / 2 : 200
        let radius = max(min(width, height) * 0.35, 60)
        let angle = (Double(index) / Double(max(count, 1))) * 2 * Double.pi
        var hash: Int32 = 0
        for scalar in id.unicodeScalars {
            hash = hash &* 31 &+ Int32(truncatingIfNeeded: scalar.value)
        }
        let phase = Double(abs(Int(hash)) % 628) / 100
        return Point(x: cx + radius * cos(angle), y: cy + radius * sin(angle), phase: phase)
    }

    /// One simulation tick, in place. `t` drives the sin/cos idle wander (same units as the
    /// web rAF timestamp: milliseconds); `wanderEnabled` gates it off under reduce-motion.
    static func step(
        points: inout [String: Point],
        ids: [String],
        edges: [Edge],
        width: Double,
        height: Double,
        t: Double,
        wanderEnabled: Bool
    ) {
        let cx = width > 0 ? width / 2 : 300
        let cy = height > 0 ? height / 2 : 200

        // Pass 1: pairwise repulsion.
        for i in 0..<ids.count {
            guard var a = points[ids[i]] else { continue }
            for j in (i + 1)..<ids.count {
                guard var b = points[ids[j]] else { continue }
                var dx = a.x - b.x
                var dy = a.y - b.y
                var d2 = dx * dx + dy * dy
                if d2 < 1 {
                    dx = Double.random(in: -0.5...0.5)
                    dy = Double.random(in: -0.5...0.5)
                    d2 = 1
                }
                let d = d2.squareRoot()
                let force = repulsion / d2
                let fx = (dx / d) * force
                let fy = (dy / d) * force
                a.vx += fx
                a.vy += fy
                b.vx -= fx
                b.vy -= fy
                points[ids[j]] = b
            }
            points[ids[i]] = a
        }

        // Pass 2: edge springs.
        for e in edges {
            guard var a = points[e.a], var b = points[e.b] else { continue }
            let rest = e.hasLabel ? springRestAI : springRestCooccur
            let dx = b.x - a.x
            let dy = b.y - a.y
            let raw = (dx * dx + dy * dy).squareRoot()
            let d = raw == 0 ? 0.01 : raw
            let force = (d - rest) * springK
            let fx = (dx / d) * force
            let fy = (dy / d) * force
            a.vx += fx
            a.vy += fy
            b.vx -= fx
            b.vy -= fy
            points[e.a] = a
            points[e.b] = b
        }

        // Pass 3: center pull + wander + damping + integrate. Fixed points just stop.
        for id in ids {
            guard var p = points[id] else { continue }
            if p.fixed {
                p.vx = 0
                p.vy = 0
                points[id] = p
                continue
            }
            p.vx += (cx - p.x) * centerK
            p.vy += (cy - p.y) * centerK
            if wanderEnabled {
                p.vx += sin(t * 0.0012 + p.phase) * wander
                p.vy += cos(t * 0.0009 + p.phase * 1.7) * wander
            }
            p.vx *= damping
            p.vy *= damping
            p.x += p.vx
            p.y += p.vy
            points[id] = p
        }
    }
}

File: android/app/src/main/java/com/nianxiang/app/ui/GraphLayout.kt
package com.nianxiang.app.ui

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/** Mutable simulation point for one graph node. Pure JVM, no Android imports, so it's
 *  unit-testable without Robolectric. */
data class GraphPoint(
    var x: Float,
    var y: Float,
    var vx: Float = 0f,
    var vy: Float = 0f,
    var fixed: Boolean = false,
    val phase: Float = 0f,
)

/** Force-directed layout for the relationship graph. Constants and the per-tick step are a
 *  line-for-line port of the web client's useForceLayout (GraphOverlay.tsx) — Android has no
 *  system reduced-motion toggle, so unlike the web version the idle wander drift always runs. */
object GraphLayout {
    const val REPULSION = 2500f
    const val SPRING_K = 0.04f
    const val SPRING_REST_AI = 90f
    const val SPRING_REST_COOCCUR = 140f
    const val CENTER_K = 0.004f
    const val DAMPING = 0.9f
    const val WANDER = 0.035f

    /** One edge for layout purposes: ids of the two endpoints + whether it carries an AI
     *  label (vs. a bare co-occurrence link), which sets the spring's rest length. */
    data class LayoutEdge(val a: String, val b: String, val hasLabel: Boolean)

    /** Initial placement for a freshly-appeared node: even spacing around a circle, phase
     *  seeded from a hash of its id (so the same node always starts at the same wander phase). */
    fun initialPosition(index: Int, total: Int, id: String, width: Float, height: Float): GraphPoint {
        val cx = if (width > 0f) width / 2f else 300f
        val cy = if (height > 0f) height / 2f else 200f
        val radius = maxOf(minOf(width, height) * 0.35f, 60f)
        var hash = 0
        for (c in id) hash = (hash * 31 + c.code)
        val angle = (index.toFloat() / maxOf(total, 1)) * (2 * Math.PI).toFloat()
        return GraphPoint(
            x = cx + radius * cos(angle),
            y = cy + radius * sin(angle),
            phase = (abs(hash) % 628) / 100f,
        )
    }

    /**
     * One simulation tick, mutating [points] in place: pairwise repulsion, edge springs
     * (rest length by [LayoutEdge.hasLabel]), center pull + idle wander drift, damping,
     * then integrate. Fixed points (being dragged) don't move.
     */
    fun step(points: Map<String, GraphPoint>, ids: List<String>, edges: List<LayoutEdge>, width: Float, height: Float, t: Float) {
        val cx = if (width > 0f) width / 2f else 300f
        val cy = if (height > 0f) height / 2f else 200f

        for (i in ids.indices) {
            val a = points[ids[i]] ?: continue
            for (j in i + 1 until ids.size) {
                val b = points[ids[j]] ?: continue
                var dx = a.x - b.x
                var dy = a.y - b.y
                var d2 = dx * dx + dy * dy
                if (d2 < 1f) {
                    dx = (Math.random() - 0.5).toFloat()
                    dy = (Math.random() - 0.5).toFloat()
                    d2 = 1f
                }
                val d = sqrt(d2)
                val force = REPULSION / d2
                val fx = (dx / d) * force
                val fy = (dy / d) * force
                a.vx += fx
                a.vy += fy
                b.vx -= fx
                b.vy -= fy
            }
        }

        for (e in edges) {
            val a = points[e.a] ?: continue
            val b = points[e.b] ?: continue
            val rest = if (e.hasLabel) SPRING_REST_AI else SPRING_REST_COOCCUR
            val dx = b.x - a.x
            val dy = b.y - a.y
            val d = sqrt(dx * dx + dy * dy).let { if (it == 0f) 0.01f else it }
            val force = (d - rest) * SPRING_K
            val fx = (dx / d) * force
            val fy = (dy / d) * force
            a.vx += fx
            a.vy += fy
            b.vx -= fx
            b.vy -= fy
        }

        for (id in ids) {
            val p = points[id] ?: continue
            if (p.fixed) {
                p.vx = 0f
                p.vy = 0f
                continue
            }
            p.vx += (cx - p.x) * CENTER_K
            p.vy += (cy - p.y) * CENTER_K
            p.vx += sin(t * 0.0012f + p.phase) * WANDER
            p.vy += cos(t * 0.0009f + p.phase * 1.7f) * WANDER
            p.vx *= DAMPING
            p.vy *= DAMPING
            p.x += p.vx
            p.y += p.vy
        }
    }
}
